numpy/_core/tests/test_stringdtype.py

```diff
@@ -701,7 +701,10 @@ def test_float_casts(typename):
     sres = np.array(inp, dtype=typename).astype("T")
     res = sres.astype(typename)
     assert_array_equal(np.array(inp, dtype=typename), res)
-    assert sres[0] == "0.1"
+    if typename == "float16":
+        assert sres[0] == "0.0999755859375"
+    else:
+        assert sres[0] == "0.1"
```

Comment:
@f2013519 interesting idea--I went with `if (val == (long long)val)` to guard the "exact" prints to "exact" integers, and it almost worked--this was the only shim I needed in the entire testsuite for things to pass locally. That probably means there is something I'm not quite getting--the StringDtype business is probably something Nathan can comment on though.

I'll go ahead and flush the CI test matrix here to see if there are any other issues reviewers would want to consider (32-bit, other platforms/architectures, etc.). Something tells me CI will find some other failures, but we'll have to wait and see.